src/Interactive/Features/Interactive/Core/InteractiveHost.RemoteService.cs

```diff
@@ -49,35 +57,36 @@ internal RemoteService(InteractiveHost host, Process process, int processId, Ser
 
             internal void HookAutoRestartEvent()
             {
-                const int ProcessExitHooked = 1;
-                const int ProcessExitHandled = 2;
-
-                int processExitHandling = 0;
+                // hook the event only once per process:
+                if (Interlocked.Exchange(ref _processExitHandling, ProcessExitHooked) == 0)
+                {
+                    Process.Exited += ProcessExitedHandler;
+                }
+            }
 
-                async void localHandler(object _, EventArgs __)
+            private async void ProcessExitedHandler(object _, EventArgs __)
+            {
+                try
                 {
-                    try
+                    if (Interlocked.Exchange(ref _processExitHandling, ProcessExitHandled) == ProcessExitHooked)
                     {
-                        if (Interlocked.Exchange(ref processExitHandling, ProcessExitHandled) == ProcessExitHooked)
-                        {
-                            Process.Exited -= localHandler;
+                        Process.Exited -= ProcessExitedHandler;
 
-                            if (!_disposing)
+                        if (!_disposing)
+                        {
+                            using (await _disposeSemaphore.DisposableWaitAsync().ConfigureAwait(false))
```

Comment:
>using (await _disposeSemaphore.DisposableWaitAsync().ConfigureAwait(false)) [](start = 28, length = 75)

I'm not sure ho is this supposed to work. Should we just signal (call ```_disposeSemaphore.Release()```) when this event is finished?